pkg/cluster/internal/providers/docker/provision.go

```diff
@@ -144,7 +144,33 @@ func commonArgs(cluster string, cfg *config.Cluster) ([]string, error) {
 		"--tty",    // allocate a tty for entrypoint logs
 		// label the node with the cluster ID
 		"--label", fmt.Sprintf("%s=%s", clusterLabelKey, cluster),
+		// user a user defined docker network so we get embedded DNS
 		"--net", fixedNetworkName,
+		// Docker supports the following restart modes:
+		// - no
+		// - on-failure[:max-retries]
+		// - unless-stopped
+		// - always
+		// https://docs.docker.com/engine/reference/commandline/run/#restart-policies---restart
+		//
+		// What we desire is:
+		// - restart on host / dockerd reboot
+		// - don't restart for any other reason
+		//
+		// This means:
+		// - no is out of the question ... it never restarts
+		// - always is a poor choice, we'll keep trying to restart nodes that were
+		// never going to work
+		// - unless-stopped will also retry failures indefinitely, similar to always
+		// except that it won't restart when the container is `docker stop`ed
+		// - on-failure is not great, we're only interested in restarting on
+		// reboots, not failures. *however* we can limit the number of retries
+		// *and* it forgets all state on dockerd restart and retries anyhow.
+		// - on-failure:0 is what we want .. restart on failures, except max
+		// retries is 0, so only restart on reboots.
+		// however this _actually_ means the same thing as always
+		// so the closest thing is on-failure:1, which will retry *once*
+		"--restart=on-failure:1",
```

Comment:
this means if the node fails the entrypoint it will actually quickly start again, which is not quite ideal, but doesn't seem too problematic.

I did quite a bit of testing around this behavior before settling on this as "ok".

Reply:
I think we have to assume several restarts for the worker nodes until the control-plane node is up, or will they keep trying?
We can't guarantee the order in which the containers are created, right?

Reply:
The entrypoint isn't aware of other nodes and ultimately becomes systemd. kubelet restarts over and over w/o exiting the container.

The containers only fail when killed or when the entry point fails prior to systemd (systemd will only exit for a signal).

So this means when a kind node crashes in the entrypoint we try to run it again once ~immediately... And it's just going to fail again.

This makes it slightly racy / slightly longer to when kind realizes the container is dead and fails / rms it unless --retain.

It will only retry *once* unless dockerd is restarted though.